In [1]:
import pandas as pd
import numpy as np

### DEALING WITH CONTENTS TABLE DATA

In [2]:
contents=pd.read_csv('Content.csv',index_col='Unnamed: 0')

In [3]:
contents.head(2)

Content ID                               User ID  \
0  97522e57-d9ab-4bd6-97bf-c24d952602d2  8d3cd87d-8a31-4935-9a4f-b319bfe05f31   
1  9f737e0a-3cdd-4d29-9d24-753f4e3be810  beb1f34e-7870-46d6-9fc7-2e12eb83ce43   

    Type        Category                                                URL  
0  photo        Studying  https://socialbuzz.cdn.com/content/storage/975...  
1  photo  healthy eating  https://socialbuzz.cdn.com/content/storage/9f7...

In [4]:
for i in contents.columns:
    print(i)

Content ID
User ID
Type
Category
URL


In [5]:
#checking for the null values
contents.isnull().sum()

Content ID      0
User ID         0
Type            0
Category        0
URL           199
dtype: int64

Here we have 199 null values in the URL column, for our analysis we don't need the URL columns, so we can drop it.

In [6]:
contents.drop(columns=['URL'],inplace=True)

In [7]:
contents.head(2)

Content ID                               User ID  \
0  97522e57-d9ab-4bd6-97bf-c24d952602d2  8d3cd87d-8a31-4935-9a4f-b319bfe05f31   
1  9f737e0a-3cdd-4d29-9d24-753f4e3be810  beb1f34e-7870-46d6-9fc7-2e12eb83ce43   

    Type        Category  
0  photo        Studying  
1  photo  healthy eating

In [8]:
#Data Exploration

In [9]:
contents['Content ID'].nunique()

1000

In [10]:
contents['User ID'].nunique()

446

In [11]:
# getting type of contents and their counts
contents['Type'].value_counts()

photo    261
video    259
GIF      244
audio    236
Name: Type, dtype: int64

In [27]:
#DEALING WITH DUPLICATES VALUES

In [12]:
# Exploring the Categories
contents['Category'].nunique()

41

In [13]:
contents['Category'].value_counts()

technology           71
animals              67
travel               67
culture              63
science              63
fitness              61
food                 61
healthy eating       61
cooking              60
soccer               58
tennis               58
education            57
dogs                 56
studying             55
veganism             48
public speaking      48
Fitness               5
Animals               4
Science               4
"soccer"              3
"culture"             3
Soccer                3
"dogs"                2
Education             2
Studying              2
Travel                2
Food                  2
"veganism"            1
"public speaking"     1
Public Speaking       1
"technology"          1
"cooking"             1
Healthy Eating        1
"studying"            1
"food"                1
Culture               1
"tennis"              1
Technology            1
"animals"             1
Veganism              1
"science"             1
Name: Category, 

In [14]:
old_values=[]
filtered_values=[]
for i in contents['Category'].unique():
    if i[0]=='"':
        old_values.append(i)
        filtered_values.append(i.replace('"',''))

In [15]:
old_values

['"culture"',
 '"studying"',
 '"animals"',
 '"soccer"',
 '"dogs"',
 '"tennis"',
 '"food"',
 '"technology"',
 '"cooking"',
 '"public speaking"',
 '"veganism"',
 '"science"']

In [16]:
filtered_values

['culture',
 'studying',
 'animals',
 'soccer',
 'dogs',
 'tennis',
 'food',
 'technology',
 'cooking',
 'public speaking',
 'veganism',
 'science']

In [17]:
#now replacing the values in the original dataframe
contents['Category'].replace(to_replace=old_values,value=filtered_values,inplace=True)

In [18]:
contents['Category'].nunique()

29

we can see that there are few letter starts with small letters and similar names starts with capital letter we have to deal with them.

In [19]:
old_values=[]
new_caps_values=[]
for i in contents['Category'].unique():
    old_values.append(i)
    new_caps_values.append(i.upper())

In [20]:
old_values

['Studying',
 'healthy eating',
 'technology',
 'food',
 'cooking',
 'dogs',
 'soccer',
 'public speaking',
 'science',
 'tennis',
 'travel',
 'fitness',
 'education',
 'studying',
 'veganism',
 'Animals',
 'animals',
 'culture',
 'Fitness',
 'Veganism',
 'Travel',
 'Education',
 'Technology',
 'Soccer',
 'Culture',
 'Food',
 'Healthy Eating',
 'Science',
 'Public Speaking']

In [21]:
new_caps_values

['STUDYING',
 'HEALTHY EATING',
 'TECHNOLOGY',
 'FOOD',
 'COOKING',
 'DOGS',
 'SOCCER',
 'PUBLIC SPEAKING',
 'SCIENCE',
 'TENNIS',
 'TRAVEL',
 'FITNESS',
 'EDUCATION',
 'STUDYING',
 'VEGANISM',
 'ANIMALS',
 'ANIMALS',
 'CULTURE',
 'FITNESS',
 'VEGANISM',
 'TRAVEL',
 'EDUCATION',
 'TECHNOLOGY',
 'SOCCER',
 'CULTURE',
 'FOOD',
 'HEALTHY EATING',
 'SCIENCE',
 'PUBLIC SPEAKING']

In [22]:
contents['Category'].replace(to_replace=old_values,value=new_caps_values,inplace=True)

In [25]:
contents['Category'].nunique()

16

In [26]:
for i in contents['Category'].unique():
    print(i)

STUDYING
HEALTHY EATING
TECHNOLOGY
FOOD
COOKING
DOGS
SOCCER
PUBLIC SPEAKING
SCIENCE
TENNIS
TRAVEL
FITNESS
EDUCATION
VEGANISM
ANIMALS
CULTURE


In [29]:
# we don't deal with user id for our project so we can drop User ID
contents.drop(columns=['User ID'],inplace=True)

In [31]:
contents.head(2)

Content ID   Type        Category
0  97522e57-d9ab-4bd6-97bf-c24d952602d2  photo        STUDYING
1  9f737e0a-3cdd-4d29-9d24-753f4e3be810  photo  HEALTHY EATING

In [32]:
#rename column name 'Type' to 'Content Type'
contents.rename(columns={'Type':'Content Type'},inplace=True)

In [33]:
contents.head(2)

Content ID Content Type        Category
0  97522e57-d9ab-4bd6-97bf-c24d952602d2        photo        STUDYING
1  9f737e0a-3cdd-4d29-9d24-753f4e3be810        photo  HEALTHY EATING

In [34]:
contents.isnull().sum()

Content ID      0
Content Type    0
Category        0
dtype: int64

FINALLY OUR WORK IS DONE FOR PREPROCESSING.

### Exploring Reaction_types dataset

In [36]:
reaction_type=pd.read_csv('ReactionTypes.csv',index_col='Unnamed: 0')

In [37]:
reaction_type.head(2)

Type Sentiment  Score
0  heart  positive     60
1   want  positive     70

In [38]:
reaction_type.shape

(16, 3)

In [39]:
reaction_type['Type'].nunique()

16

In [76]:
reaction_type.rename(columns={'Type':'Reaction Type'},inplace=True)

In [40]:
#We have 16 Different types of reactions

In [41]:
reaction_type

Type Sentiment  Score
0         heart  positive     60
1          want  positive     70
2       disgust  negative      0
3          hate  negative      5
4    interested  positive     30
5   indifferent   neutral     20
6          love  positive     65
7    super love  positive     75
8       cherish  positive     70
9         adore  positive     72
10         like  positive     50
11      dislike  negative     10
12    intrigued  positive     45
13      peeking   neutral     35
14       scared  negative     15
15      worried  negative     12

### Exploring the Reaction dataset

In [50]:
reaction=pd.read_csv('Reactions.csv',index_col='Unnamed: 0')

In [51]:
reaction.head(2)

Content ID                               User ID  \
0  97522e57-d9ab-4bd6-97bf-c24d952602d2                                   NaN   
1  97522e57-d9ab-4bd6-97bf-c24d952602d2  5d454588-283d-459d-915d-c48a2cb4c27f   

      Type             Datetime  
0      NaN  2021-04-22 15:17:15  
1  disgust  2020-11-07 09:43:50

In [52]:
reaction.shape

(25553, 4)

In [53]:
for i in reaction.columns:
    print(i)

Content ID
User ID
Type
Datetime


In [54]:
# WE DON'T NEED THE USER ID COLUMN, DROP IT
reaction.drop(columns=['User ID'],inplace=True)

In [57]:
# Dealing with the null values
reaction.isnull().sum()

Content ID      0
Type          980
Datetime        0
dtype: int64

In [27]:
reaction['Type'].nunique()

16

Here reaction Type is a key data point, if there is null values exist then we have to drop it.

In [58]:
# Here we have to drop null values in the column 'Type'  
reaction.dropna(inplace=True)

In [59]:
reaction.shape

(24573, 3)

In [60]:
reaction.isnull().sum()

Content ID    0
Type          0
Datetime      0
dtype: int64

In [62]:
# Rename the column 'Type' to 'Reation Type'
reaction.rename(columns={'Type':'Reaction Type'},inplace=True)

In [63]:
reaction

Content ID Reaction Type             Datetime
1      97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust  2020-11-07 09:43:50
2      97522e57-d9ab-4bd6-97bf-c24d952602d2       dislike  2021-06-17 12:22:51
3      97522e57-d9ab-4bd6-97bf-c24d952602d2        scared  2021-04-18 05:13:58
4      97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust  2021-01-06 19:13:01
5      97522e57-d9ab-4bd6-97bf-c24d952602d2    interested  2020-08-23 12:25:58
...                                     ...           ...                  ...
25548  75d6b589-7fae-4a6d-b0d0-752845150e56       dislike  2020-06-27 09:46:48
25549  75d6b589-7fae-4a6d-b0d0-752845150e56     intrigued  2021-02-16 17:17:02
25550  75d6b589-7fae-4a6d-b0d0-752845150e56    interested  2020-09-12 03:54:58
25551  75d6b589-7fae-4a6d-b0d0-752845150e56       worried  2020-11-04 20:08:31
25552  75d6b589-7fae-4a6d-b0d0-752845150e56       cherish  2021-01-04 04:55:11

[24573 rows x 3 columns]

## Data Merging

In [66]:
contents.head(2)

Content ID Content Type        Category
0  97522e57-d9ab-4bd6-97bf-c24d952602d2        photo        STUDYING
1  9f737e0a-3cdd-4d29-9d24-753f4e3be810        photo  HEALTHY EATING

In [68]:
contents.shape

(1000, 3)

In [72]:
reaction['Content ID'].nunique()

962

In [69]:
reaction.head(2)

Content ID Reaction Type             Datetime
1  97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust  2020-11-07 09:43:50
2  97522e57-d9ab-4bd6-97bf-c24d952602d2       dislike  2021-06-17 12:22:51

In [70]:
reaction.shape

(24573, 3)

In [73]:
# Merging the reation and contents tables
final_data=pd.merge(reaction,contents,how='inner',on='Content ID')

In [74]:
final_data

Content ID Reaction Type  \
0      97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust   
1      97522e57-d9ab-4bd6-97bf-c24d952602d2       dislike   
2      97522e57-d9ab-4bd6-97bf-c24d952602d2        scared   
3      97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust   
4      97522e57-d9ab-4bd6-97bf-c24d952602d2    interested   
...                                     ...           ...   
24568  75d6b589-7fae-4a6d-b0d0-752845150e56       dislike   
24569  75d6b589-7fae-4a6d-b0d0-752845150e56     intrigued   
24570  75d6b589-7fae-4a6d-b0d0-752845150e56    interested   
24571  75d6b589-7fae-4a6d-b0d0-752845150e56       worried   
24572  75d6b589-7fae-4a6d-b0d0-752845150e56       cherish   

                  Datetime Content Type    Category  
0      2020-11-07 09:43:50        photo    STUDYING  
1      2021-06-17 12:22:51        photo    STUDYING  
2      2021-04-18 05:13:58        photo    STUDYING  
3      2021-01-06 19:13:01        photo    STUDYING  
4      2020-08-23 12:25:58        photo    STUDYING  
...                    ...          ...         ...  
24568  2020-06-27 09:46:48        audio  TECHNOLOGY  
24569  2021-02-16 17:17:02        audio  TECHNOLOGY  
24570  2020-09-12 03:54:58        audio  TECHNOLOGY  
24571  2020-11-04 20:08:31        audio  TECHNOLOGY  
24572  2021-01-04 04:55:11        audio  TECHNOLOGY  

[24573 rows x 5 columns]

In [88]:
reaction_type

Reaction Type Sentiment  Score
0          heart  positive     60
1           want  positive     70
2        disgust  negative      0
3           hate  negative      5
4     interested  positive     30
5    indifferent   neutral     20
6           love  positive     65
7     super love  positive     75
8        cherish  positive     70
9          adore  positive     72
10          like  positive     50
11       dislike  negative     10
12     intrigued  positive     45
13       peeking   neutral     35
14        scared  negative     15
15       worried  negative     12

In [79]:
#merging final_data and reaction type table
final_data=pd.merge(final_data,reaction_type,how='inner',on='Reaction Type')

In [80]:
final_data

Content ID Reaction Type  \
0      97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust   
1      97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust   
2      97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust   
3      9f737e0a-3cdd-4d29-9d24-753f4e3be810       disgust   
4      230c4e4d-70c3-461d-b42c-ec09396efb3f       disgust   
...                                     ...           ...   
24568  435007a5-6261-4d8b-b0a4-55fdc189754b         adore   
24569  435007a5-6261-4d8b-b0a4-55fdc189754b         adore   
24570  4e4c9690-c013-4ee7-9e66-943d8cbd27b7         adore   
24571  4e4c9690-c013-4ee7-9e66-943d8cbd27b7         adore   
24572  4e4c9690-c013-4ee7-9e66-943d8cbd27b7         adore   

                  Datetime Content Type        Category Sentiment  Score  
0      2020-11-07 09:43:50        photo        STUDYING  negative      0  
1      2021-01-06 19:13:01        photo        STUDYING  negative      0  
2      2021-04-09 02:46:20        photo        STUDYING  negative      0  
3      2021-03-28 21:15:26        photo  HEALTHY EATING  negative      0  
4      2020-08-04 05:40:33        photo  HEALTHY EATING  negative      0  
...                    ...          ...             ...       ...    ...  
24568  2020-10-04 22:26:33        audio        VEGANISM  positive     72  
24569  2020-09-18 10:50:50        audio        VEGANISM  positive     72  
24570  2020-10-31 03:58:44          GIF         CULTURE  positive     72  
24571  2020-06-25 15:12:29          GIF         CULTURE  positive     72  
24572  2020-12-17 16:32:57          GIF         CULTURE  positive     72  

[24573 rows x 7 columns]

In [81]:
final_data.isnull().sum()

Content ID       0
Reaction Type    0
Datetime         0
Content Type     0
Category         0
Sentiment        0
Score            0
dtype: int64

In [82]:
final_data.to_csv('Final_Processed_Data.csv')

In [87]:
final_data.tail(5)

Content ID Reaction Type  \
24568  435007a5-6261-4d8b-b0a4-55fdc189754b         adore   
24569  435007a5-6261-4d8b-b0a4-55fdc189754b         adore   
24570  4e4c9690-c013-4ee7-9e66-943d8cbd27b7         adore   
24571  4e4c9690-c013-4ee7-9e66-943d8cbd27b7         adore   
24572  4e4c9690-c013-4ee7-9e66-943d8cbd27b7         adore   

                  Datetime Content Type  Category Sentiment  Score  
24568  2020-10-04 22:26:33        audio  VEGANISM  positive     72  
24569  2020-09-18 10:50:50        audio  VEGANISM  positive     72  
24570  2020-10-31 03:58:44          GIF   CULTURE  positive     72  
24571  2020-06-25 15:12:29          GIF   CULTURE  positive     72  
24572  2020-12-17 16:32:57          GIF   CULTURE  positive     72

In [84]:
grp_by_category=final_data.groupby('Category')
grp_by_category.groups

{'ANIMALS': [64, 67, 198, 199, 224, 225, 226, 241, 242, 243, 244, 247, 248, 249, 250, 251, 294, 295, 325, 328, 329, 339, 340, 341, 398, 399, 400, 403, 404, 405, 406, 407, 408, 441, 448, 469, 505, 506, 507, 508, 509, 510, 522, 523, 524, 525, 526, 556, 557, 558, 559, 614, 615, 635, 636, 637, 638, 639, 694, 708, 709, 710, 711, 719, 720, 741, 750, 767, 927, 928, 938, 939, 940, 945, 946, 981, 986, 987, 988, 1012, 1013, 1014, 1033, 1045, 1046, 1047, 1062, 1063, 1064, 1079, 1080, 1101, 1156, 1167, 1200, 1212, 1217, 1222, 1273, 1274, ...], 'COOKING': [54, 55, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 215, 216, 217, 218, 219, 255, 256, 287, 288, 289, 313, 314, 337, 338, 393, 394, 409, 423, 424, 425, 426, 427, 428, 429, 439, 442, 443, 444, 481, 482, 492, 493, 494, 495, 571, 572, 619, 669, 670, 671, 743, 791, 828, 829, 830, 831, 832, 833, 879, 956, 957, 996, 999, 1000, 1001, 1002, 1030, 1031, 1032, 1034, 1035, 1042, 1043, 1044, 1070, 1089, 1090, 1100, 1120, 1124, 1136, 1137, 1138, 1139, 1

In [91]:
Top_5_Categories=grp_by_category['Score'].sum()

In [97]:
Top_5_Categories=Top_5_Categories.sort_values(ascending=False).head(5)

In [98]:
Top_5_Categories

Category
ANIMALS           74965
SCIENCE           71168
HEALTHY EATING    69339
TECHNOLOGY        68738
FOOD              66676
Name: Score, dtype: int64

In [99]:
Top_5_Categories.to_csv('Top_5_Categories.csv')